src/utils/flux/connect/connect.js

```diff
@@ -0,0 +1,62 @@
+import React from 'react';
+
+/**
+ * Higher order component
+ *
+ * Takes a variadic number of arguments where the leading arguments are
+ * Stores and the the tail of those arguments is a function mapping from
+ * store states to the props of the composed component.
+ *
+ * Usage like so:
+ *
+ *    function mapStateToProps(stateA, stateB) {
+ *      return {
+ *        propFromA: stateA.get('someProp'),
+ *        propFromB: stateB.get('someProp')
+ *      }
+ *    }
+ *
+ *    const ConnectedComponent = connect(storeA, storeB, mapStateToProps)(TargetComponent)
```

Comment:
This looks really good, trying to think if we can introduce it as a smaller breaking change. For example we could follow the existing API and add the `mapStateToProps` function as the third arg:

```
connect(TargetComponent, [storeA, storeB], (stateA, stateB) => {
  return {
    foo: stateA.get('foo'),
    bar: stateB.get('bar')
  };
});
```

Then we could replace the existing function more easily - I'd rather avoid having two different versions of it.

In terms of what will break:

* Requires third arg (we could default it to the entire store state, but this would give devs a lazy way out - would rather they're forced to write the function and think about what props their component needs?)
* Changing `this.state.x` to `this.props.x`.
* Tests? Not sure what the fallout here would be exactly.

Reply:
I think it would make more sense to make this a replacement for the existing connect method and deprecate the old. There shouldn't be any need for the old connect and it's use should be actively discouraged.

Also, functions in the following format play a lot nicer with component composition:
`const createHOC = (...config) => (TargetComponent) => { /* ... */ }`

See [recompose](https://github.com/acdlite/recompose) and its `compose` function. For example:

```
const withStoreState = connect(storeA, storeB, mapStateToProps)

const MyComposedComponent = compose(
  withStoreState,
  // ...
)(MyComponent)
```

Reply:
Sure, what I meant was deprecating the old one - but I was trying to ask if we can minimise the amount of changes we'd need with removing uses of the old one.

We could add a deprecation message to the previous version and allow access to this one, and this may give some time for developers to convert old ones and to start using the new one - then with version 3 or 4 we can remove it.

Reply:
Ah, makes sense!

I think its worth breaking the API in the long term and being consistent going forward with the HOC pattern mentioned above, otherwise you end up with the following boilerplate every time you want to use component composition in friendly way:

```
const withStoreState = (Component) => connect(Component, [storeA, storeB], mapStateToProps)

const MyComposedComponent = compose(
  withStoreState
  // ...
)(MyComponent)
```